src/webgpu/shader/types.ts

```diff
@@ -204,10 +216,34 @@ export function* generateTypes({
   containerType: ContainerType;
   /** Whether to wrap the baseType in `atomic<>`. */
   isAtomic?: boolean;
-}) {
+}): Generator<
+  {
+    type: string; // WGSL name for the generated type
+    _kTypeInfo: {
+      // WGSL name for:
+      // - the generated type if it is scalar or atomic
+      // - the column vector type if the generated type is a matrix
+      // - the base type if the generated type is an array
+      elementBaseType: string;
+      // Layout details if host-shareable, and undefined otherwise.
+      layout: undefined | AlignmentAndSize;
+      supportsAtomics: boolean;
+      // The number of elementBaseType items in the container.
+      arrayLength: number;
+      // 0 for scalar and vector.
+      // For a matrix type, this is the number of rows in the matrix.
+      innerLength?: number;
+    };
+  },
+  void
+> {
   const scalarInfo = kScalarTypeInfo[baseType];
   if (isAtomic) {
     assert(scalarInfo.supportsAtomics, 'type does not support atomics');
+    assert(
+      containerType === 'scalar' || containerType === 'array',
+      'atomics are only supported for scalars and arrays'
```

Comment:
nit: I was confused about this for a minute because it sounded like `atomic<array<f32>>`
```suggestion
      "can only generate atomic inner types with containerType 'scalar' or 'array'"
```